### PR TITLE
Add Possibilities to typechecker

### PR DESCRIPTION
**Pull request by amosr**
See issue #81.

### CODE REVIEW

**Comment by jystic** [comment]
:potable_water: looks good, although I have to admit I don't really understand the meat of it because I'm not too familiar with the type checker

**Comment by amosr** [comment]
@tranma I expect we'll have some conflicts around case expressions and their possibilities. do you think I should wait for you to merge before merging this, or just go ahead and we'll deal with it when it happens?

**Comment by tranma** [comment]
@amosr not sure. We could try and see if the case expression branch is good to merge today?

**Comment by tranma** [comment]
woah, documentation :open_mouth:

File: test/cli/repl/t20-lexer/expected [hunk 1]
```diff
@@ -23,17 +23,14 @@ ok, loaded test/cli/repl/data.psv, 8 rows
...
+      max                  : Element a -> Aggregate Possibly a
+      mean                 : (Num a) => Element a -> Aggregate Double
       date                 : Element Definitely DateTime
-      min                  : Element a b -> Aggregate a b
-      sd                   : (Element =: ReturnOfLet Pure t$137
-                             ,t$135 =: ReturnOfLet a a
-                             ,t$136 =: ReturnOfLet a t$135
-                             ,t$137 =: ReturnOfLet a t$136) => a b Double -> Aggregate b Double
+      min                  : Element a -> Aggregate Possibly a
+      sd                   : (Element =: ReturnOfLet Pure a) => a Double -> Aggregate Double
```
Comment (tranma) [inline test/cli/repl/t20-lexer/expected]: nice

**Comment by amosr** [comment]
> woah, documentation :open_mouth:
does that count as approval-to-merge?

File: src/Icicle/Source/Type/Compounds.hs [hunk 2]
```diff
@@ -60,6 +62,23 @@ freeT t
...
+                    p q r   -> Set.unions
+                             $ fmap freeT
+                             [ p, q, r ]
+    CReturnOfLatest p q r   -> Set.unions
+                             $ fmap freeT
+                             [ p, q, r ]
+    CPossibilityJoin p q r  -> Set.unions
+                             $ fmap freeT
+                             [ p, q, r ]
+
```
Comment (tranma) [inline src/Icicle/Source/Type/Compounds.hs]: probably get a "pattern not matched" here on the temporality constraints

**Comment by tranma** [comment]
looks alright to me, I mostly read the doc and glazed over the implementation though :eyeglasses:

File: src/Icicle/Source/Type/Compounds.hs (hunk 2 above)
Comment (amosr) [inline src/Icicle/Source/Type/Compounds.hs]: doh…

File: test/cli/repl/t20-lexer/expected (hunk 1 above)
Comment (jystic) [inline test/cli/repl/t20-lexer/expected]: That's pretty nice

Comment (amosr) [inline test/cli/repl/t20-lexer/expected]: not as nice as it should be…

Comment (tranma) [inline test/cli/repl/t20-lexer/expected]: I guess ideally we don't even want to see `ReturnOfLet`, does `Element =: ReturnOfLet Pure a` just implies `a` is `Element`?

Comment (amosr) [inline test/cli/repl/t20-lexer/expected]: yeah, it does really. I'm tempted to remove the ReturnOfLet stuff altogether..

**Pull request merged**